### PR TITLE
Two-Tile Stream-K

### PR DESCRIPTION
**Pull request by AlexBrownAMD**
Implement the two-tile stream-k algorithm. Enabled by setting StreamK=3.

### CODE REVIEW

File: Tensile/KernelWriterAssembly.py [hunk 1]
```diff
@@ -3652,13 +3659,39 @@ def openPersistentLoop(self, kernel):
...
+          # iter count after all extra iters have been distributed
+          kStr += inst("s_mul_i32", sgpr("StreamKIter"), sgpr("StreamKIdx"), sgpr("SKItersPerWG"), "StreamK starting iteration (case: after extra iters)")
+          kStr += inst("s_add_u32", sgpr("StreamKIter"), sgpr("StreamKIter"), sgpr("skExtraIters"), "Add extra iters")
+          kStr += inst("s_add_u32", sgpr("StreamKIterEnd"), sgpr("StreamKIter"), sgpr("SKItersPerWG"), "StreamK ending iteration (case: after extra iters)")
+          # iter count before all extra iters have been distributed
+          # stmp+1 = SKItersPerWG + 1 extra iteration
+          sIter = self.sgprPool.checkOut(2, "SKIter", preventOverflow=0)
+          kStr += inst("s_add_u32", sgpr(sIter+1), sgpr("SKItersPerWG"), 1, "Spread out extra iterations")
+          kStr += inst("s_mul_i32", sgpr(sIter), sgpr("StreamKIdx"), sgpr(sIter+1), "StreamK starting iteration (case: before extra iters)")
+          kStr += inst("s_add_u32", sgpr(sIter+1), sgpr(sIter), sgpr(sIter+1), "StreanJ endung iteration (case: before extra iters)")
```
[the remark this answers is not in the document]
Comment (AlexBrownAMD) [reply Tensile/KernelWriterAssembly.py]: Fixed

File: Tensile/KernelWriterAssembly.py [hunk 2]
```diff
@@ -3652,13 +3659,39 @@ def openPersistentLoop(self, kernel):
...
         kStr += inst("s_mov_b32", sgpr("StreamKIdx"), sgpr("WorkGroup0"), "Save original StreamK index")
-        kStr += inst("s_mul_i32", sgpr("StreamKIter"), sgpr("WorkGroup0"), sgpr("ItersPerWG"), "StreaK starting iteration")
-        kStr += inst("s_add_u32", sgpr("StreamKIterEnd"), sgpr("StreamKIter"), sgpr("ItersPerWG"), "StreamK ending iteration")
-        kStr += inst("s_min_u32", sgpr("StreamKIterEnd"), sgpr("StreamKIterEnd"), sgpr("TotalIters"), "Cap ending iter at total iters")
-        kStr += inst("s_cmp_lt_u32", sgpr("StreamKIter"), sgpr("StreamKIterEnd"), "Make sure there's work to do")
-        kStr += self.longBranchScc0("label_%04u" % (self.getLabelNum("KernelEnd")), positiveOnly=True)
-        # kStr += inst("s_cbranch_scc0", "label_%04u" % (self.getLabelNum("KernelEnd")), "edge case that work doesn't divide well")        
-        kStr += self.undefineSgpr("TotalIters")
+        if kernel["StreamK"] < 3: # Basic SK
+          kStr += inst("s_mul_i32", sgpr("StreamKIter"), sgpr("StreamKIdx"), sgpr("SKItersPerWG"), "StreaK starting iteration")
```
[the remark this answers is not in the document]
Comment (AlexBrownAMD) [reply Tensile/KernelWriterAssembly.py]: Fixed

**Comment by AlexBrownAMD** [comment]
> By the way, using MIWorkGroup=[2,1] causes hang with StreamK= 2 or 3. Like this: MatrixInstruction: [32, 32, 8, 1, 1, 2,1, 2,1]

Yea, it requires MIWorkGroup[0] * MIWorkGroup[1] == 4 for now. I've added a reject case to SolutionStructs to check for this.

File: Tensile/Source/lib/source/ContractionSolution.cpp [hunk 3]
```diff
@@ -610,16 +612,35 @@ namespace Tensile
...
+                if(sizeMapping.streamK < 3) // Basic SK
+                {
+                    uint32_t itersPerWave = CeilDivide(totalIters, rv.numWorkGroups.x);
+                    rv.args.append<uint32_t>("SKItersPerWG", itersPerWave);
+                }
+                else if(sizeMapping.streamK == 3) // Two-tile SK
+                {
+                    uint32_t numOutputTiles = tiles;
+                    bool     bigEnough      = numOutputTiles > skGrid;
+                    uint32_t skTiles
```
[the remark this answers is not in the document]
Comment (nakajee) [reply Tensile/Source/lib/source/ContractionSolution.cpp:635]: Thanks for your update.
But, I am still quite not sure why we use numOutputTiles % skGrid here.
Could you explain why we need to add numOutputTiles % skGrid here?

[the remark this answers is not in the document]
Comment (AlexBrownAMD) [reply Tensile/Source/lib/source/ContractionSolution.cpp:635]: For sure. numOutputTiles is total number of output tiles and skGrid is number of workgroups we launch. skGrid defaults to the number of CUs for now, but we've been experimenting with launching different size grids, like 2 workgroups per CU or other configurations.

The two-tile algorithm does some stream-k iterations up front, and then an even number of data-parallel tiles afterwards. So, the number of data-parallel tiles must be an even multiple of skGrid.

(numOutputTiles % skGrid) makes sure that any extra tiles that won't divide evenly across the CUs are processed as stream-k tiles instead.

[the remark this answers is not in the document]
Comment (nakajee) [reply Tensile/Source/lib/source/ContractionSolution.cpp:635]: I dumped kernel arguments and spent time to understand this part.
numOutputTiles  is split into 2 parts like this?
  1) dp loop(?):  numOutputTiles - skTile
  2) sk loop(?):  skTile (= skGrid + numOutputTiles % skGrid)
Probably, it is better to add some more explanation about the tile allocation logic (probably in Common.py?).

**Review by nakajee — APPROVED** [review]
I am OK with the current change.
Please make sure all tests pass with the change.

**Pull request merged**